tiledb/ml/data_apis/tensorflow_sparse.py

```diff
@@ -0,0 +1,154 @@
+            )
+
+        # Get number of observations
+        rows = x_array.schema.domain.shape[0]
+
+        # Get x and y shapes
+        cls.x_shape = (None,) + x_array.schema.domain.shape[1:]
+        cls.y_shape = (None,) + y_array.schema.domain.shape[1:]
+
+        # Get x and y data types
+        cls.x_dtype = x_array.schema.attr(0).dtype
+        cls.y_dtype = y_array.schema.attr(0).dtype
+
+        if isinstance(y_array, tiledb.SparseArray):
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_sparse,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.SparseTensorSpec(shape=cls.y_shape, dtype=cls.y_dtype),
+                ),
+                args=(rows, batch_size),
+            )
+        else:
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_dense,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.TensorSpec(shape=cls.y_shape, dtype=cls.y_dtype),
+                ),
+                args=(rows, batch_size),
+            )
+
+    @classmethod
+    def _generator_sparse_sparse(cls, rows: int, batch_size: int) -> tuple:
+        """
+        A generator function that yields the next training batch.
+        :param x: TileDB array. An opened TileDB array which contains features.
+        :param y: TileDB array. An opened TileDB array which contains labels.
+        :param rows: Integer. The number of observations in x, y datasets.
+        :param batch_size: Integer. Size of batch, i.e., number of rows returned per call.
+        :return: Tuple. Tuple that contains x and y batches.
+        """
+        # Loop over batches
+        # https://github.com/tensorflow/tensorflow/issues/44565
+        for offset in range(0, rows, batch_size):
+            y_batch = cls.y[offset : offset + batch_size]
+            x_batch = cls.x[offset : offset + batch_size]
+
+            # TODO: Both for dense case support multiple attributes
+            values_y = list(y_batch.items())[0][1]
+            values_x = list(x_batch.items())[0][1]
+
+            # Transform to TF COO format y data
+            y_data = np.array(values_y).flatten()
+            y_coords = []
+            for i in range(0, cls.y.schema.domain.ndim):
+                dim_name = cls.y.schema.domain.dim(i).name
+                y_coords.append(np.array(y_batch[dim_name]))
+
+            # Transform to TF COO format x data
+            x_coords = []
+            for i in range(0, cls.x.schema.domain.ndim):
+                dim_name = cls.x.schema.domain.dim(i).name
+                x_coords.append(np.array(x_batch[dim_name]))
+
+            x_data = np.array(values_x).flatten()
+
+            if x_data.shape[0] != y_data.shape[0]:
+                raise Exception(
```

Comment:
1. Perhaps compute `x_data` and `_y_data` first and fail fast if this condition fails, before computing `y_coords`/`x_coords`
2. Exception is the base of (most) exceptions; better use a more specific one (e.g. `ValueError`)

Reply:
1. You right on that. We can get a slight performance gain especially in big arrays. I will fix it.
2. Perfect I will specialise it!